server/controllers/dashboardController/index.js

```diff
@@ -2,36 +2,54 @@ import { ObjectId } from 'mongodb'
 import UserModel from '../../models/UserModel'
 import ConfigModel from '../../models/ConfigModel'
 import DashboardService from '../../services/DashboardService'
+import DashboardDataProcessor from '../../data_processors/DashboardDataProcessor'
 
 const DashboardsController = {
   show: async (req, res) => {
     try {
       const { appDb } = req.app.locals
       const { study, subject } = req.params
-
       const user = await UserModel.findOne(appDb, { uid: req.user.uid })
       const userConfigurationQuery = {
         _id: new ObjectId(user.preferences.config),
       }
       const config = await ConfigModel.findOne(appDb, userConfigurationQuery)
       const flatConfig = Object.values(config.config).flat()
 
-      const dashboardService = new DashboardService(
+      const { dashboardDataCursor, consentDate } = new DashboardService(
         appDb,
         study,
         subject,
         flatConfig
       )
-      const { matrixData, consentDate } =
-        await dashboardService.createDashboard()
+      const participantConsentDate = await consentDate()
+      const dataStream = await dashboardDataCursor()
+      const participantDataMap = new Map()
+
+      dataStream.on('data', (doc) => {
+        const dayData = doc?.dayData.length ? doc.dayData : []
+
+        participantDataMap.set(doc.assessment, dayData)
+      })
+
+      await new Promise((resolve, reject) => {
+        dataStream.on('end', () => resolve())
+
+        dataStream.on('error', (err) => reject(err))
+      })
+
+      const dashboardProcessor = new DashboardDataProcessor(
+        flatConfig,
+        participantDataMap
+      )
```

Comment:
can we move this logic out of the controller and into the service?

Reply:
maybe expose a method on the service for `participantDataMap`? Then we can `await service.participantDataMap()`?

Reply:
[62415d4](https://github.com/AMP-SCZ/dpdash/pull/630/commits/62415d47393169a2b7d7e879b26a8ca7ade93d83)

Reply:
That suggestion makes sense thanks for that